### ISSUE TITLE
Specify implementation sequencing and acceptance checks

### ISSUE THREAD

**Issue by komaksym** [komaksym/kbc-solution#9]
## Question

What implementation sequence, notebook checkpoints, test strategy, runtime and reproducibility contract, and artifact-validation gates should govern building the final Python Jupyter notebook and its generated outputs? Define the smallest milestone plan that verifies the fixed data contract and leakage controls early, keeps the nested validation and bootstrap workload tractable and deterministic, executes cleanly from a fresh kernel, and proves that the exactly two-page executive-summary PDF and both CSV exports reconcile with the executed notebook.

**Comment by komaksym** [comment]
## Resolution

### Governing implementation rule

Build the smallest statistically honest, interview-ready solution in one visible Python Jupyter notebook. Stop for review after each milestone. Do not introduce a hidden source package, a separate test architecture, persistent model caches, or an optimization project. Preserve the fixed analytical contract unless a later decision explicitly supersedes it.

When background research conflicts with later closed decisions, the later decisions govern: use the classifier-only 5 outer folds × 3 fixed seeds design, 3 inner folds, and the 300+300 default uncertainty design from [Choose model selection and policy-validation rules](https://github.com/komaksym/kbc-solution/issues/6), subject only to the runtime fallback below.

### Five reviewable milestones

```text
M1 source contract
  -> M2 semantic features + logistic policy baseline
    -> M3 nested model ladder + held-out policy evidence
      -> M4 locked refit + stability + CSV exports
        -> M5 two-page PDF + clean full execution + final reconciliation
```

1. **Source contract and environment:** create the pinned environment, configuration cell, source lineage, safe keyed joins, population masks, fixed row-count/schema assertions, and anomaly inventory. Stop after proving the 1,615/969/646 populations, the 28 unknown activity records split 18/10, and all leakage exclusions.
2. **Semantic preprocessing and baseline:** implement structural product absence, observed-zero preservation, unknown sex/activity handling, fold-local imputation/encoding/scaling, the compact feature registry, and the complete three-product all-logistic expected-revenue policy. Stop after a fresh-kernel smoke execution proves the end-to-end baseline path.
3. **Model selection and policy validation:** add the fixed LR/RF/HGB ladder, gated XGBoost, feature comparisons, sigmoid calibration, 5×3 nested client folds, proper-score guardrails, policy promotion gate, baselines, tail sensitivity, and held-out diagnostics. Stop after all outer OOF rows reconcile exactly and no outer-fold information enters selection.
4. **Locked refit, uncertainty, and exports:** lock the selected architecture, refit on all 969 labeled clients, score all 646 targets, run uncertainty/stability resampling, allocate at most 96 positive-value contacts, and write both CSVs. Stop after deterministic ranking and exact cross-file reconciliation pass.
5. **Executive artifact and final acceptance:** generate the exactly two-page PDF, execute the complete notebook from a fresh kernel in full mode, verify every artifact and runtime gate, and retain only full-mode evidence in the submitted notebook and outputs.

Each milestone is a separately reviewable implementation slice; do not begin the next milestone until the current slice is accepted.

### Notebook, environment, and paths

- Keep analytical logic, compact helper functions, docstrings, tests/assertions, results, and interpretation visible in one self-contained notebook. Assertions are executable notebook tests; do not create a hidden `src/` package or separate pytest suite for this assignment.
- Use a minimal `pyproject.toml` plus `uv.lock`, including the complete deterministic dependency set required by the fixed ladder, including XGBoost even though its challenger is gated. Do not install packages dynamically inside the notebook.
- Record Python and package versions, fixed seeds, source SHA-256, run mode, bootstrap counts, elapsed runtime, and the reference-machine description in the notebook.
- Use one configuration cell with an explicit portable `WORKBOOK_PATH` and fixed `OUTPUT_DIR`; provide a clear missing-file error. Never hard-code the current Downloads path, search the filesystem automatically, or create timestamped output directories.

### Smoke/full execution and runtime

- Provide one code path with `smoke` and `full` modes. Smoke mode exercises all three products, preprocessing, selection, calibration, policy allocation, exports, and artifact checks with one fixed outer seed and 20 historical-value plus 20 locked-refit bootstrap runs. Smoke results are development evidence only and must never appear in the final PDF or CSVs.
- Full mode uses the fixed 5 outer folds × 3 seeds procedure and defaults to 300 paired historical-value bootstrap resamples plus 300 locked-model target-stability refits.
- Use fixed random states and one controlled parallel layer; avoid nested oversubscription and persistent caches.
- Excluding the one-time environment installation, a clean full-mode run has a hard gate of 15 minutes on the recorded modern reference laptop.
- Do not start an optimization effort if the run exceeds 15 minutes, and do not reduce folds, seeds, candidate families, hyperparameter grids, calibration, or policy checks. Keep the 300 historical-value resamples because they operate on fixed OOF records and are cheap. Reduce only the expensive locked-model stability refits from 300 to 100; if the same clean run still exceeds 15 minutes, reduce them to 50. Record and disclose the final count because frequency resolution becomes 1% at 100 runs and 2% at 50 runs.

### Bootstrap semantics

- For historical policy-value uncertainty, resample complete clients while preserving all three product outcomes. Apply bootstrap multiplicities as weights to the fixed leakage-safe OOF score/outcome records, recompute the exact one-offer capacity policy, and never refit or reselect models inside these 300 paired resamples.
- For target stability, refit only the locked architecture on client bootstrap samples and score the fixed 646-client targeting population.
- `offer_assignment_frequency_<offer>` is the unconditional share of all locked refits in which the client is both selected and assigned that offer. The three offer frequencies must sum to `target_selection_frequency` for every target client. These are stability diagnostics, not purchase probabilities.

### PDF and validation cadence

- Generate the executive PDF with Matplotlib `PdfPages` as exactly two fixed-size pages. Assert the PDF page count is exactly two and render both pages during final QA so clipping or overflow cannot pass merely because the file has two page objects.
- After each executable milestone, run its narrow assertions and a fresh-kernel smoke execution. Run one fresh-kernel full execution only for final acceptance.
- Milestone gates fail fast on source schema/count/key/join drift; fabricated labels; predictor leakage; non-fold-local learned preprocessing; incomplete OOF coverage; invalid probabilities or scores; nondeterministic allocation; output-schema drift; duplicate/missing target clients; more than 96 contacts; offer-frequency identities; CSV disagreement; PDF page/render failure; stale smoke artifacts; or runtime above the declared gate after the bootstrap fallback.
- Final acceptance requires an executed notebook with no failed cells, exactly 646 unique rows in `target_scores.csv`, an exact selected subset of at most 96 unique clients in `illustrative_contact_list.csv`, one offer per selected client, deterministic top-positive-value ordering, an exactly two-page rendered PDF, and exact reconciliation of every headline number in the PDF with notebook tables and exported CSVs.

No implementation-sequencing, reproducibility, runtime, testing, or artifact-acceptance decisions remain open.
